doc/source/dev/core-dev/distributing.rst.inc

```diff
@@ -52,35 +53,53 @@ here.  For details, see https://scipy.github.io/devdocs/building/.
 dependencies reported by projects.  Because SciPy gets regular bug reports
 about this, we go in a bit of detail here.
 
 SciPy only reports its dependency on NumPy via ``install_requires`` if NumPy
-isn't installed at all on a system.  This will only change when there are
-either 32-bit and 64-bit Windows wheels for NumPy on PyPI or when
-``pip upgrade`` becomes available (with sane behavior, unlike ``pip install
--U``, see `this PR
-<https://github.com/pypa/pip/pull/3194>`_).  For more details, see
-`this summary <https://mail.python.org/pipermail/distutils-sig/2015-October/027161.html>`_.
-
-The situation with ``setup_requires`` is even worse; pip_ doesn't handle that
-keyword at all, while ``setuptools`` has issues (here's a `current one
-<https://bitbucket.org/pypa/setuptools/issues/391>`_) and invokes
-``easy_install`` which comes with its own set of problems (note that SciPy doesn't
-support ``easy_install`` at all anymore; issues specific to it will be closed
-as "wontfix").
+isn't installed at all on a system, or when building wheels with
+``bdist_wheel``. SciPy no longer uses ``setup_requires`` (which in the past
+invoked ``easy_install``); build dependencies are now handled only via
+``pyproject.toml``. ``pyproject.toml`` relies on PEP 517; ``pip`` has
+``--no-use-pep517`` and ``--no-build-isolation`` flags that may ignore
+``pyproject.toml`` or treat it differently - if users use those flags, they
+are responsible for installing the correct build dependencies themselves.
+
+
+.. _numpy-version-ranges:
+
+Version ranges for NumPy and other dependencies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For dependencies it's important to set lower and upper bounds on their
+versions. For *build-time* dependencies, they are specified in
+``pyproject.toml`` and the versions will _only_ apply to the SciPy build
+itself. It's fine to specify either a range or a specific version for a
+dependency like ``wheel`` or ``setuptools``. For NumPy we have to worry
+about ABI compatibility too, hence we specify the version with ``==``
+to the lowest supported version (because NumPy's ABI is backward but not
+forward compatible).
+
+For *run-time dependencies* (currently only ``numpy``), we specify the range
+of versions in ``install_requires`` in ``setup.py``. Getting the upper bound
+right is slightly tricky.  If we don't set any bound, a too-new version
+will be pulled in a few years down the line, and NumPy may have deprecated and
+removed some API that SciPy depended on by then. On the other hand if we set
+the upper bound to the newest already-released version, then as soon as a new
+NumPy version is released there will be no matching SciPy version that works
+with it. Given that NumPy and SciPy both release in a 6-monthly cadence and
+that features that get deprecated in NumPy should stay around for another two
+releases, we specify the upper bound as ``<1.xx+3.0`` (where ``xx`` is the
+minor version of the latest already-released NumPy.
```

Comment:
According to [NEP 23](https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals) deprecation cycles are two releases, so should this be `1.[xx + 2]` instead?

Reply:
I got it wrong in the first version of this PR, I think `<xx + 3` is correct. Example:

- NumPy is at 1.20.1
- Scipy 1.7.x branches. There are zero deprecation warnings coming from NumPy in the Scipy test suite.
- NumPy 1.21.0 deprecates a function
- Released SciPy 1.7.0 will start showing those deprecation warnings if it uses that function
- NumPy will only remove that function in 1.23.0 (which would break SciPy 1.7.0)

Hence if numpy is at 1.20.x now, the requirement should be `< 1.23.0`. (All `1.22.x` are fine).

Of course numpy can create unintented breakage, but hopefully that will then be fixed in a next bugfix release.